Tar 197 - Extend generator to create non-shared views in apps

lukehansell-hx opened the pull request:
#### What does this PR do?
It would be nice to be able to easily build the non shared React views inside our apps (that are based on the webapp-template). To this end I've extended the generator so that it checks for a `package.json` file in any parent directory, if it finds one then it assumes that we're inside a webapp-template project and begins the process of building an internal non-shared view. If it doesn't find a `package.json` file then it begins the process of building an external shared view.

#### What unit or integration tests does this PR have?
I've rewritten the original tests as they now only apply to the external generator. I've added new tests for the master generator and the internal one.

#### What selenium tests does this PR have?
none.

#### How should a developer review this?
Check over the code to make sure I've not left any daft things in there and that the code makes sense and is legible.

#### How should this be manually tested?
After checking out this repo you will have to run `npm link` inside it to make it globally accessible to the yeoman generator. You will also have to have `yo` and `browserify` installed globally.

You may run into an error EMFILE. This is due to Browserify attempting to open too many files in parallel. Adding `ulimit -n 2560` to your bash_profile should resolve this.

1. checkout the webapp-temaplte repo and run the generator inside it using `yo ui-component` and check the internal generator was used.
  - you should be asked which feature to build for
  - you should see the new view in the feature folder chosen.
2. run the generator using `yo ui-component` in a directory without a parent dir containing a package.json file and check the external generator was used.
  - you should be asked for a description of your view
  - your current directory should have the new view in it (including dev and script directories and a whole lot more!)
  - running `npm run build-dev && open dev/index.html` should load your default browser and if you look the source you should see an element with the class of `ui-component-[your view name]`
3. run `yo ui-component:external` anywhere and check the external generator is run (see above)
4. attempting to run `yo ui-component:internal` outside of a project or in a project with no `code` dir in the root should throw an error

#### Any background context you want to provide?
The React views built by this generator are intended for use in our webapp-template based apps that we're beginning to build.

There's recently been some name changing, so webapp-views are now ui-components. This is to bring this work in line with the css toolkit the UXUI team are working on.

#### What are the relevant tickets?
[TAR-197](https://holidayextras.jira.com/browse/TAR-197)

#### Screenshots (if appropriate)

#### What gif best describes how you feel about this work?
![](http://media.giphy.com/media/YxMACiNIvfSc8/giphy.gif)
---
#### Quality checklist (for PR author to complete BEFORE code review):
- [x] I've checked there is appropriate unit test coverage.
- [x] I've checked there is appropriate selenium test coverage.
- [x] I've included regression tests for any bug fixes.
- [x] I've updated documentation with any changes to procedures.
- [x] I've tested this on mobile Android, Safari, Chrome for any visual changes.
- [x] I've tested this on desktop IE9/10/11, Firefox, Chrome for any visual changes.
- [x] I've checked any significant visual changes with a member of the UX/UI pod.
- [x] I've checked this work against the requirements of the Jira.
- [x] I've added and updated translations for all supported languages
- [x] I've informed those who need to know of any unforeseen impact of this work (database migrations, deployment procedure changes).
- [x] I am ready for this to be code reviewed, merged and tested.
- [x] I've thought about writing a blog post about this

#### IP or Developer Review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] +1 from me!

#### Software Engineer or Developer review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] +1 from me!

#### Software Engineer or project guru review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!

LouisPryer commented:
Picking up IP/Dev Review. Just looking over this repo so I can better understand myself with it before I review this :)

LouisPryer commented:
Not entirely sure how to test this, I am not sure if I have access to the webapp-template repo since I can't find it in Github or Bitbucket :/

lukehansell-hx commented:
@LouisPryer Can you access it here: https://github.com/holidayextras/webapp-template?

LouisPryer commented:
Nope, sorry I don't seem to have access

lukehansell-hx commented:
I'm closing this PR for the time being as it requires @michaelcarter's PR to be merged first https://github.com/holidayextras/webapp-template/pull/4

lukehansell-hx commented:
reopening now that Mike's is merged. The internal generator will fail if not run inside the webapp-template project.

dancork commented:
taking dev review

dancork commented:
Manually tested and worked fine.

dancork commented on generators/app/index.js:
```diff
@@ -1,118 +1,17 @@
+    var dir;
+
+    dir = findParentDir.sync(path.resolve(''), 'package.json');
+    if (dir) {
```
As dir is scoped here and only used one, could we simplify to...

```
initializing: function () {
  if(findParentDir.sync(path.resolve(''), 'package.json')){
    this.composeWith('webapp-view:internal');
  } else {
    this.composeWith('webapp-view:external');
  }
```

dancork commented on generators/external/index.js:
```diff
@@ -0,0 +1,122 @@
+    this.fs.copyTpl(
+      this.templatePath('dev/example.jsx'),
+      this.destinationPath('dev/example.jsx'),
+      {name: this.options.camelName.charAt(0).toUpperCase() + this.options.camelName.slice(1)}
```
We do this a few times. Could make it a helper function to keep code clean.

Also, why don't we do this in writingCodeDir method?

dancork commented on generators/external/index.js:
```diff
@@ -0,0 +1,122 @@
+var generators = require('yeoman-generator');
+var path = require('path');
+
+module.exports = generators.Base.extend({
```
There are some similarities between this generator and the internal. Is it worth creating our own `base` generator and then extending that instead?

dancork commented:
I think it's worth updating the readme. I found the installation instructions on here - relating to using npm link - more helpful. Especial as several people will be working from their dev copy.

Also think it's worth mentioning, though it may seem obvious, how to install yeoman. I just went on the website and used `npm install -g yo bower grunt-cli gulp`.

lukehansell-hx commented:
@dancork I've addressed some of your feedback and have added a base class which all functionality (which can be) is extended from.
I believe this makes the generators look a bit cleaner. If you could take a look again I'd appreciate it :smile:

I've also added some error checking which should provide more assistance when attempting to run the generator not in the expected location.

I'll update the README as a separate JIRA.

lukehansell-hx commented:
I've now merged in master which had the name change to generator-ui-component and have done some conflict resolution for this.

michaelcarter commented:
Shotgun SE/Guru

michaelcarter commented on README.md:
```diff
@@ -40,8 +60,17 @@ The generator is built to work with Yeoman. Run the following command to install
 
+If you would like to specifically run the internal generator you can specify:
+
+`yo webapp-view:internal`
```
I think the language of the generator needs to be consistent with the more generic `ui-component` naming we settled on last week, so we can do `yo ui-component:internal` and `yo ui-component:external`. We'll need to update the language in the template project to reference these as `ui components` rather than shared views too. Naming is fun...

dancork commented:
Looks cleaner now. Still working as expected. :+1:

michaelcarter commented on README.md:
````diff
@@ -27,9 +32,24 @@ The structure created looks as follows:
+
+```
+(./code/products/views/)
+├ [module-name]/
````
`component-name`? We were worried module made people associate these things too closely with node modules, rather than ui modules, so we settled on components.

dancork commented:
Readme does need updating though. One of the PR tick boxes is `I've updated documentation with any changes to procedures.`

michaelcarter commented:
The `npm run build-dev` and `npm run build-dist` tasks aren't working, but they're not in master either.

michaelcarter commented:
Just the build-dist and build-dev tasks not working now then I'm all good to +1

lukehansell-hx commented:
@michaelcarter what output are you getting when you try to run the tasks?

lukehansell-hx commented:
I think I've figured out your issue @michaelcarter. The command line argument for the variant of the redirectify command read `--dir $VARIANT` which if the VARIANT value was blank would translate to `--dir  ` which redirectify would interpret as the --dir flag on it's own so would be parsed as `--dir=true`. If you try to pass this into the `path.resolve()` function it would quite rightly fail.

To resolve this issue I've added quotation marks `""` around the variant flag so that it now reads `--dir "$VARIANT"`. If no variant is now provided this resolves to `--dir ""` or `--dir=""`. Much better. 

Can you let me know if this resolves the problem you were seeing?

GeorgeF72 commented:
I'll take the Ip/Dev review as Louis isn't in to finish it.

GeorgeF72 commented:
All good :+1:

lukehansell-hx commented:
@dancork, @michaelcarter any chance you guys could take another look at this today please?

michaelcarter commented:
@lukehansell-hx Yes, will take a look once we're finished with the IP group tasks.

michaelcarter commented:
`npm run build-dist` is still broken.

michaelcarter commented:
+1

dancork commented:
:+1:
